### ISSUE TITLE
AVL Tree bug

### ISSUE THREAD

**Issue by prankymat** [hollance/swift-algorithm-club#61]
AVL Tree will crash with following conditions:

```swift
let tree = AVLTree<Int, String>()

tree.insert(1, "five")
tree.insert(2, "four")
tree.insert(3, "three")
tree.insert(4, "two")
tree.insert(5, "one")

tree.delete(5)
tree.delete(1)
tree.delete(4) // Crashes here
tree.delete(2)
tree.delete(3)
```

There are other combinations of remove that will crash. Examples:
```swift
[2, 3, 1, 5, 4]  // Crashes at 3
[4, 5, 3, 2, 1]  // Crashes at 5
[3, 2, 5, 4, 1]  // Crashes at 2
```

Swift was trying to unwrap a nil reference in ```delete```, causing the program to crash.

Please advice thanks!

**Comment by hollance** [comment]
This does not surprise me. The AVL Tree implementation was written by someone new to Swift (if I recall correctly) and may still have some issues.

If you're an expert on AVL trees, feel free to take a look at the code and improve it. In any case, it's near the top of my to-do list of things to improve.

**Comment by mitghi** [comment]
I've checked it, you are right. There seems to be a bug in `delete()`. The problem is caused by a wrong increment to balance factor. I test it and try to fix the bug and will report here. 
Thanks

**Comment by prankymat** [comment]
Please refer fix to #81

**Comment by hollance** [comment]
W00t. I added the above test case to the tests.
